### ISSUE TITLE
[v4.0] Possibility to render multiple filtered areas

### ISSUE THREAD

**Issue by andre-tronin** [nelmio/NelmioApiDocBundle#1754]
Users of our API get access only for some routes, which have been unlocked for that user.
For example:
User1 can use Route1 and Route3
User2 can use Route2 and Route3
It can be changed later, so the list of visible areas must be dynamic.
If the user can actually use the route, there is an authorization. So the problem is only the visibility of areas documentation for the user.
For the v3 we had build a workaround:
```php
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\DependencyInjection\ServiceLocator;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Security\Core\Security;
use Symfony\Component\Routing\Annotation\Route;

/**
 * @Route("/customer/docs", name="app_customer_api-docs", methods={"GET"})
 */
class SwaggerUiController extends AbstractController
{
    private ApiDocGeneratorFacade $apiDocGeneratorFacade;
    private Security $security;

    public function __construct(ApiDocGeneratorFacade $apiDocGeneratorFacade, Security $security)
    {
        $this->apiDocGeneratorFacade = $apiDocGeneratorFacade;
        $this->security = $security;
    }

    public function __invoke(): Response
    {
        return $this->render(
            '@NelmioApiDoc/SwaggerUi/index.html.twig',
            ['swagger_data' => ['spec' => $this->apiDocGeneratorFacade->generate($this->security->getUser())]]
        );
    }
}

class ApiDocGeneratorFacade
{
    private ServiceLocator $apiDocGeneratorLocator;

    public function __construct(ServiceLocator $apiDocGeneratorLocator)
    {
        $this->apiDocGeneratorLocator = $apiDocGeneratorLocator;
    }

    public function generate(User $apiUser): array
    {
        $areas = Areas::getAreasByApiUser($apiUser);

        $apiDocGenerators = array_map(fn (string $area) => $this->apiDocGeneratorLocator->get($area), $areas);

        $specs = [];
        foreach ($apiDocGenerators as $apiDocGenerator) {
            $specs = array_replace_recursive($specs, json_decode($apiDocGenerator->generate()->toJson(), true));
        }

        return $specs;
    }
}
```
nelmio_api_doc.yml:
```yaml
nelmio_api_doc:
    documentation:
        info:
            title: api
            description: api
            version: 2.2.2
        components:
            securitySchemes:
                Basic:
                    type: http
                    scheme: basic
                Bearer:
                    type: apiKey
                    name: Authorization
                    in: header
    areas:
        default:
            path_patterns:
                - '^/api'
        area_1:
            name_patterns:
               - '^route_1'
        area_2:
            name_patterns:
               - '^route_2'
        area_3:
            name_patterns:
               - '^route_3'
```
But with update to v4 it stopped to work:
`
User Notice: $ref "#/components/schemas/SomeModel" not found for @OA\Items() in \App\Domain\Model\SomeModelCollection->getItems in /path/src/Domain/Model/SomeModelCollection.php
`
This error occurs in the case when two or more areas, that has to be shown uses same `@Model` somewhere.
If there are no such areas, the workaround works as expected.

What should we change, that the workaround always works?
Or perhaps there is another solution for our needs that we did not discover jet?

**Comment by GuilhemN** [comment]
I don't really see why your code would fail. Aren't you using manually defined references somewhere?

If not, could you provide a small reproducer so that I can better understand your error?

**Comment by andre-tronin** [comment]
@GuilhemN  thank you for quick answer!
I tried to make a minimalistic example

Controllers:
```php
use Nelmio\ApiDocBundle\Annotation\Model;
use OpenApi\Annotations as OA;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\Routing\Annotation\Route;

/**
 * @Route("/api/route-1", name="route_1", methods={"GET"}, format="json")
 */
class Route1Controller extends AbstractController
{
    /**
     * @OA\Get(
     *      tags={"route_1"},
     *      summary="Returns a list of some models",
     *      description="Returns a list of some models",
     *      @OA\Response(
     *          description="Returns a list of some models.",
     *          response=200,
     *          @Model(type=App\Domain\Model\SomeModelCollection::class, groups={"api"})
     *      )
     * )
     */
    public function __invoke()
    {
    }
}
/**
 * @Route("/api/route-2", name="route_2", methods={"GET"}, format="json")
 */
class Route2Controller extends AbstractController
{
    /**
     * @OA\Get(
     *      tags={"route_2"},
     *      summary="Returns a list of some models",
     *      description="Returns a list of some models",
     *      @OA\Response(
     *          description="Returns a list of some models.",
     *          response=200,
     *          @Model(type=App\Domain\Model\SomeModelCollection::class, groups={"api"})
     *      )
     * )
     */
    public function __invoke()
    {
    }
}
```
Models:
```php
namespace App\Domain\Model;

use Nelmio\ApiDocBundle\Annotation\Model;
use OpenApi\Annotations as OA;

class SomeModelCollection
{
    /**
     * @OA\Property(type="array", @OA\Items(ref=@Model(type=App\Domain\Model\SomeModel::class)))
     */
    public array $items;
}
class SomeModel
{

}
```
ApiDocGeneratorFacade:
```php
class ApiDocGeneratorFacade
{
    private ServiceLocator $apiDocGeneratorLocator;

    public function __construct(ServiceLocator $apiDocGeneratorLocator)
    {
        $this->apiDocGeneratorLocator = $apiDocGeneratorLocator;
    }

    public function generate(User $apiUser): array
    {
        $areas = [
            'area_1',
            'area_2',
        ];

        $apiDocGenerators = array_map(fn (string $area) => $this->apiDocGeneratorLocator->get($area), $areas);

        $specs = [];
        foreach ($apiDocGenerators as $apiDocGenerator) {
            $specs = array_replace_recursive($specs, json_decode($apiDocGenerator->generate()->toJson(), true));
        }

        return $specs;
    }
}
```
**Error**:
```
User Notice: $ref "#/components/schemas/SomeModel" not found for @OA\Items() in \Doctrine\Common\Annotations\DocParser->Annotation() in /path/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/DocParser.php 
```
If you change
```php
  $areas = [
            'area_1',
            'area_2',
        ];
```
to

```php
  $areas = [
            'area_1',
        ];
```
or to

```php
  $areas = [
            'area_2',
        ];
```
docs will be generated as expected without any errors

**Comment by GuilhemN** [comment]
Sorry I can't reproduce your issue...

Are all you libraries up to date? What version of NelmioApiDocBundle and zircote/swagger-php are you using?

**Comment by andre-tronin** [comment]
I think I found why there is no error in the example I provided.
PHPDoc of $items in SomeModelCollection lack @Groups("api") annotation (use Symfony\Component\Serializer\Annotation\Groups;)

I've just made an empty symfony 5 project with only this example. And as I first could not reproduce it too. But than I saw, that the Model was actually ignored, because there was no Groups annotation. After I added it, the error occurred.

nelmio/api-doc-bundle v4.0.1
zircote/swagger-php 3.1.0

**Comment by andre-tronin** [comment]
@GuilhemN could you reproduce the error? Any updates here?

**Comment by GuilhemN** [comment]
Well it still works fine for me, I'm not sure what I've done differently from you...

I created a reproducer at https://github.com/GuilhemN/nelmio-issue-1754-reproducer where I tried to reproduce the conditions you described but I still can't reproduce your issue.
Could you have a look at it?
Or maybe directly share the simple reproducer you created where this happens?

**Comment by andre-tronin** [comment]
Hello @GuilhemN, by compairing and adding libraries from our project into reproducer I found the library, that makes problems.
That's doctrine/dbal
I created a pull request, so you can test it. Or you can add it by yourself with "composer require doctrine/dbal"

**Comment by GuilhemN** [comment]
Ok in fact this is was fixed in part in https://github.com/nelmio/NelmioApiDocBundle/pull/1632 but an occurrence was forgotten, see https://github.com/nelmio/NelmioApiDocBundle/pull/1769/files#diff-5dc7d7dcf78698e0d0cf5b1113d5c44c036a217a689f35f3d336bacae13d06abR49.

Could you check whether #1769 does correctly fix your issue please? (it fixes the reproducer at least)

**Comment by andre-tronin** [comment]
I don't know, how I correctly test with a branch.
However, when I changed "annotation_reader" to "annotations.reader" in DependencyInjection/Compiler/ConfigurationPass.php and Resources/config/services.xml as in #1769 , the issue was fixed.
Thank you! I'll wait now for release, so I can update to nelmio 4.x

**Comment by andre-tronin** [comment]
the issue is fixed in v4.2.0
